nums/core/grid/grid.py

```diff
@@ -105,64 +106,40 @@ def nbytes(self):
         return nbytes
 
 
-class DeviceID:
-    @classmethod
-    def from_str(cls, s: str):
-        a, b = s.split("/")
-        node_id, node_addr = a.split("=")
-        device_type, device_id = b.split(":")
-        return DeviceID(int(node_id), node_addr, device_type, int(device_id))
-
-    def __init__(self, node_id: int, node_addr: str, device_type: str, device_id: int):
-        self.node_id: int = node_id
-        self.node_addr: str = node_addr
-        self.device_type: str = device_type
-        self.device_id: int = device_id
-
-    def __str__(self):
-        return self.__repr__()
-
-    def __hash__(self):
-        return hash(self.__repr__())
-
-    def __repr__(self):
-        return "%s=%s/%s:%s" % (
-            self.node_id,
-            self.node_addr,
-            self.device_type,
-            self.device_id,
-        )
-
-    def __eq__(self, other):
```

Comment:
Can we define this in the dataclass? Should be defined as `self.node_id == other.node_id and self.device_id == other.device_id`. We do several thousand of these comparisons in the optimizer branch (which we will merge under experimental package soon) and I think a quick comparison of these integers will likely be faster than what dataclass may do by default (e.g. I expect perhaps it compares all properties).

If you think dataclass imp is faster, can you point to relevant documentation explaining how it compares equality by default?